inst/include/dust/gpu/rng.hpp

```diff
@@ -0,0 +1,144 @@
+#ifndef DUST_GPU_RNG_HPP
+#define DUST_GPU_RNG_HPP
+
+#include <dust/gpu/xoshiro.hpp>
+#include <dust/gpu/distr/binomial.hpp>
+#include <dust/gpu/distr/normal.hpp>
+#include <dust/gpu/distr/poisson.hpp>
+#include <dust/gpu/distr/uniform.hpp>
+
+namespace dust {
+
+struct RNGptr {
+  uint64_t* state_ptr;
+  unsigned int state_stride;
+  unsigned int particle_stride;
+};
+
+// Read state from global memory
+template <typename T>
+__device__
+rng_state_t<T> loadRNG(RNGptr& rng_state, int p_idx) {
+  rng_state_t<T> state;
+  for (int i = 0; i < XOSHIRO_WIDTH; i++) {
+    state.s[i] = rng_state.state_ptr[p_idx * rng_state.particle_stride +
+                                      i * rng_state.state_stride];
+  }
+  return state;
+}
+
+// Write state into global memory
+template <typename T>
+__device__
+void putRNG(rng_state_t<T>& rng, RNGptr& rng_state, int p_idx) {
+  for (int i = 0; i < XOSHIRO_WIDTH; i++) {
+    rng_state.state_ptr[p_idx * rng_state.particle_stride +
+                        i * rng_state.state_stride] = rng.s[i];
+  }
+}
+
+template <typename real_t>
+class pRNG { // # nocov
+public:
+  pRNG(const size_t n, const uint64_t seed) {
+    dust::Xoshiro rng(seed);
+    for (int i = 0; i < n; i++) {
+      _rngs.push_back(rng);
+      rng.jump();
+    }
+
+    CUDA_CALL(cudaMalloc((void** )&_d_rng_state.state_ptr,
+                          n * XOSHIRO_WIDTH * sizeof(uint64_t)));
+    // Set to be interleaved
+    _d_rng_state.state_stride = n;
+    _d_rng_state.particle_stride = 1;
+    put_state_device();
+  }
+
+  ~pRNG() {
+    CUDA_CALL(cudaFree(_d_rng_state.state_ptr));
+  }
+
+  RNGptr state_ptr() { return _d_rng_state; }
+
+  size_t size() const {
+    return _rngs.size();
+  }
+
+  void jump() {
+    get_state_device();
+    for (size_t i = 0; i < _rngs.size(); ++i) {
+      _rngs[i].jump();
+    }
+    put_state_device();
+  }
+
+  void long_jump() {
+    get_state_device();
+    for (size_t i = 0; i < _rngs.size(); ++i) {
+      _rngs[i].long_jump();
+    }
+    put_state_device();
+  }
+
+  std::vector<uint64_t> export_state() {
+    std::vector<uint64_t> state;
+    const size_t n = XOSHIRO_WIDTH; // TODO: nicer in cpu version
+    state.reserve(size() * n);
+    for (size_t i = 0; i < size(); ++i) {
+      uint64_t * s = _rngs[i].get_rng_state();
+      for (size_t j = 0; j < n; ++j) {
+        state.push_back(s[j]);
+      }
+    }
+    return state;
+  }
+
+private:
+  // delete move and copy to avoid accidentally using them
+  pRNG ( const pRNG & ) = delete;
+  pRNG ( pRNG && ) = delete;
+
+  void put_state_device() {
+    std::vector<uint64_t> interleaved_state(size() * XOSHIRO_WIDTH);
```

Comment:
Change variable name to `flattened_state` (as it is not necessarily interleaved)

Reply:
done in b8ffc38